servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,97 +337,175 @@ public void onComplete() {
+                    // wrapper. There is no more concurrency because neither `itemsTerminated` nor `accumulate` will be
+                    // invoked again.
+                    @SuppressWarnings("unchecked")
+                    final ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(
+                            ((NextAccumulatorHolder<T, B>) cMaybeAccumulator).accumulator, terminalNotification);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                        bCancellable.cancel();
+                        return;
+                    }
+                } else if (cMaybeAccumulator == null) {
+                    if (maybeAccumulatorUpdater.compareAndSet(this, null, TERMINATED)) {
+                        // Terminated without any produced items or boundaries
+                        try {
+                            bCancellable.cancel();
+                        } finally {
+                            terminateTarget(null, target, terminalNotification);
+                        }
+                        return;
+                    }
+                } else {
+                    @SuppressWarnings("unchecked")
+                    final CountingAccumulator<T, B> accumulator = (CountingAccumulator<T, B>) cMaybeAccumulator;
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator,
+                            new ItemsTerminated<>(accumulator, terminalNotification))) {
+                        try {
+                            bCancellable.cancel();
+                        } finally {
+                            terminateIfPossible(accumulator, target, terminalNotification);
+                        }
+                        return;
+                    }
                 }
             }
         }
 
         void boundariesTerminated(final Throwable cause, final Subscriber<?> target) {
             maybeAccumulator = TERMINATED;
-            target.onError(cause);
+            safeOnError(target, cause);
         }
 
-        private static <T, B> void terminateTarget(@Nullable final Accumulator<T, B> accumulator,
-                                                   final Subscriber<? super B> target,
-                                                   final TerminalNotification terminalNotification,
-                                                   final Cancellable bCancellable) {
-            try {
-                if (accumulator != null) {
-                    try {
-                        target.onNext(accumulator.finish());
-                    } catch (Throwable t) {
-                        safeOnError(target, t);
-                        return;
-                    }
+        private <T, B> void deliverOnNext(final Accumulator<T, B> accumulator, final Subscriber<? super B> target) {
+            // an alternative of `decrementAndGet` without update of the special states:
```

Comment:
did you consider adjusting the states so we can just use `decrementAndGet` as a light weight alternative? For example would using `-1` to indicate "needs more demand", `-2` would be "got demand and delivered" (only set here), and `-3+` be terminated work? worth consideration since this is invoked on each `onNext`.

Reply:
The only difference between `accumulateAndGet` and `decrementAndGet` is the ternary operator for `prev <= LAST_SPECIAL_STATE` check which I believe is not a bottleneck.
Let me keep it as-is. Don't want to mess with another state transitioning in my head and it will be safer to use `Long.MIN_VALUE` range for the special states to avoid situations when boundaries do not respect demand and may deliver more than requested.